pytorch_multi_things_2.py

```diff
@@ -0,0 +1,221 @@
+
+    def train_batch(self, batch):
+        input, target = batch
+
+        # We deal with zero_grad() (do we?)
+
+        with autocast():
+            output0 = self.model0(input)
+            output1 = self.model1(input)
+            loss0 = loss_fn(2 * output0 + 3 * output1, target)
+            loss1 = loss_fn(3 * output0 - 5 * output1, target)
+
+        self.scaler.scale(loss0).backward(retain_graph=True)
+        self.scaler.scale(loss1).backward()
+
+        # Users clip gradients in callbacks specified in the scaler.step() operation
+        # (this is because we handle gradient aggregation ourselves)
+        def my_clip_grads_0(...):
+            return ...
+
+        def my_clip_grads_1(...):
+            return ...
+
+        self.scaler.step(self.optimizer0, clip_grads=my_clip_grads_0)
+        self.scaler.step(self.optimizer1, clip_grads=my_clip_grads_1)
+
+        # We handle scaler.update()
+
+
+"""
+Example 3: a proposal using the existing apex.amp mixed precision.
+
+we should be basically API compatible with the GradScaler, even if we are not using it.  That way
+when we do want to support it, it's easy, and also we are not making up an API totally from-scratch.
+"""
+
+
+class MyPytorchTrial(PyTorchTrial):
+    def __init__(self, context):
+        self.context = context
+        mod0 = self.context.model(my_make_model())
+        mod1 = self.context.model(my_make_model())
+        opt0 = self.context.optimizer(my_make_optimizer(mod0.parameters()))
+        opt1 = self.context.optimizer(my_make_optimizer(mod1.parameters()))
+
+        # This should work fine with apex.amp, and has the added benefit that you no longer have
+        # to configure amp via the experiment config... which in turn means you could do a
+        # hyperparameter search over it, if you wanted to.
+        (self.model0, self.model1), (self.optimizer0, self.optimizer1) = \
+            self.context.amp_init([mod0, mod1], [opt0, opt1]
+
+        # suppose the user wants one lr_scheduler auto-stepped every epoch...
+        self.lr_scheduler0 = context.lr_scheduler(
+            my_make_lr_scheduler(self.optimizer0), step_mode=STEP_EVERY_EPOCH
+        )
+
+        # and one scheduler stepped totally manually
+        self.lr_scheduler1 = context.lr_scheduler(
+            my_make_lr_scheduler(self.optimizer0), step_mode=MANUAL_STEP
+        )
+
+    def train_batch(self, batch):
+        input, target = batch
+
+        output0 = self.model0(input)
+        output1 = self.model1(input)
+        loss0 = loss_fn(2 * output0 + 3 * output1, target)
+        loss1 = loss_fn(3 * output0 - 5 * output1, target)
+
+        # TODO: name this better, but it is analagous to the scaler.scale() function
+        self.context.loss_tensor(loss0).backward(retain_graph=True)
+        self.context.loss_tensor(loss1).backward(retain_graph=True)
+
+        def my_clip_grads_0(...):
+            return ...
```

Comment:
@rb-determined-ai I don't like this strategy very much, but I think it is the best way we can do gradient clipping (if we are also doing gradient aggregation.

@aaron276h I don't like this, and I think calling a single gradient callback with a name argument corresponding to each optimizer would be better.